src/goto-instrument/contracts/contracts.cpp

```diff
@@ -920,23 +1121,69 @@ void code_contractst::apply_loop_contract(
       if(inner == outer)
         continue;
 
-      if(loop_nesting_graph[outer].loop.contains(
-           loop_nesting_graph[inner].target))
-        loop_nesting_graph.add_edge(outer, inner);
+      if(loop_nesting_graph[outer].content.contains(
+           loop_nesting_graph[inner].head_target))
+      {
+        if(!loop_nesting_graph[outer].content.contains(
+             loop_nesting_graph[inner].end_target))
+        {
+          log.error()
+            << "Overlapping loops at:\n"
+            << loop_nesting_graph[outer].head_target->source_location()
+            << "\nand\n"
+            << loop_nesting_graph[inner].head_target->source_location()
+            << "\nLoops must be nested or sequential for contracts to be "
+               "enforced."
+            << messaget::eom;
+        }
+        loop_nesting_graph.add_edge(inner, outer);
+      }
     }
   }
 
+  // make sure all children of a contractified loop also have contracts
+  while(!to_check_contracts_on_children.empty())
+  {
+    const auto loop_idx = to_check_contracts_on_children.front();
+    to_check_contracts_on_children.pop_front();
+
+    const auto &loop_node = loop_nesting_graph[loop_idx];
+    if(
+      loop_node.assigns_clause.is_nil() && loop_node.invariant.is_nil() &&
+      loop_node.decreases_clause.is_nil())
+    {
+      log.error()
+        << "Inner loop at: " << loop_node.head_target->source_location()
+        << " does not have contracts, but an enclosing loop does.\n"
+        << "Please provide contracts for this loop, or unwind it first."
+        << messaget::eom;
+      throw 0;
+    }
+
+    for(const auto child_idx : loop_nesting_graph.get_predecessors(loop_idx))
+      to_check_contracts_on_children.push_back(child_idx);
+  }
+
   // Iterate over the (natural) loops in the function, in topo-sorted order,
   // and apply any loop contracts that we find.
   for(const auto &idx : loop_nesting_graph.topsort())
   {
     const auto &loop_node = loop_nesting_graph[idx];
+    if(
```

Comment:
Does this correspond to a case where we have a loop without contract ? What happens if this loop is itself embedded in an outer loop which has a contract ? 

Would this result in applying the loop contract on the outer loop but without ensuring that the body of the outer loop is loop-free ?

Reply:
> Considering the check performed at line 1151 is this case really possible ?

Yes. Line 1151 checks that all loops within an outer loop with contracts also have contracts. We throw an error and stop instrumentation if we find a single violation.

This line skips instrumentation for the (outer) loops that don't have contracts.

> Does this correspond to a case where we have a loop without contract ? What happens if this loop is itself embedded in an outer loop which has a contract ?

1. Yes
2. It would have been checked at line 1151

> Would this result in applying the loop contract on the outer loop but without ensuring that the body of the outer loop is loop-free ?

I'm not sure how that would be possible.